### DISCUSSION TITLE
auto heading + graph view

### DISCUSSION THREAD

**Discussion by RazyRo** [reorproject/reor#200] [Ideas]
as Related notes are connected automatically via vector similarity so if we can use LLM-powered RAG on the corpus of notes to create/suggest Automatic titles/subtitles for notes saves time also keeps everything organized, especially when you have a lots of notes if we allow the AskFile Feature to rewrite the current file or suggest Automatic titles/subtitles for notes. The titles/subtitles also give a quick idea of what each part of your notes is about, helping you understand better and easy to connect related topics/subtopics in your notes. | Graph view - In Reor, notes are divided into chunks and stored in an internal vector database. the graphic view function in Obsidian which connect by [[]] to surround the phrase or text: [[link text]]. When typing the brackets, Obsidian offers autocomplete suggestions, which are based on the already existing note title or keywords but have to link manually by [[]] and visually in graph view section https://help.obsidian.md/Plugins/Graph+view. Check out this Obsidian plugin (https://github.com/AlexW00/obsidian-note-linker) for a similar concept, where it suggests note references based on existing titles or keywords. i think notes can be automatically connected based on their similarity, by enabling an LLM RAG system we can suggest links between points or notes (which auto heading might also help to connect). This eliminates the need for manual searching through keywords for linking.

**Comment by samlhuillier** [comment]
Thank you for your suggestions! 

On automatic titling, how would you see that actually implemented? I think this could be implemented with an in-editor LLM interface which would let you highlight parts of the note and tell the llm something like "Generate a title and subtitles".

On linking, we have two big things coming in the next release (i.e. this week): Obsidian-like square bracket linking & a feature that let's you highlight any block of text and show the semantic links corresponding to only that block. We think that'll be quite a powerful workflow, giving you the ability to leverage vector search to then generate Obsidian like backlinks...What do you think?

**Reply by RazyRo** [reply]
Let's reverse the solution to the problem. 

If we think of a knowledge management system, it's knowledge—collecting, organising, and sharing it easily. This system is like a central hub for creating, collaborating, and finding information. (For personal use, sharing and collaborating are not needed.)

 I like Reor because it's great at using LLM for knowledge creation. As for me, less is more. 

A knowledge graph helps represent knowledge more naturally, making it easier to see connections and get new ideas. While I use Obsidian, I find it lacking as my notes increase—I struggle to organise and link them the way I want manually.

questions:
How do we connect notes?
-We can use something like square brackets in Obsidian to manually connect related parts.

What if we have a lot of notes?
- Maybe we can use the LLM to help us.

Can the LLM help with exact linking?
- It might be able to.

Here's what I think:
We can make tags by clicking on the generate tags button of our notes and letting the language model suggest tags for us under each paragraph. as well we can select which tags fit best.
Or
we can set up a system where, after highlighting, the LLM suggests keywords for us to turn into custom tags.


But there's a challenge:
if we use inductive tagging, we might end up with too many tags. If we use deductive tagging, we need to have a set of rules and tags beforehand.

So, why not combine both? We could start with a set of tags and then adjust them as we go along, like how Notably.ai (https://www.notably.ai/blog/introducing-automatic-highlights-and-tagging-to-notably) suggests tags based on only that block.

This way, we can use the LLMs suggestions Tags related to - text blocks ( inductive tagging) + tags that have already been used and can be related to that block (deductive tagging). 

it's easier for inductive tagging because relates to highlighted text blocks but It might be harder to collect tags that have already been used and processed for highlighted text blocks to see if they can be related to that block.

This might refine our tagging process by leveraging vector search to help in deductive as well as inductive tagging.

**Reply by RazyRo** [reply]
some repos i found i don't the best semantic knowledge graph there is but as basic search - https://github.com/rahulnyk/knowledge_graph
https://github.com/Alibaba-NLP/KB-NER
https://github.com/zjunlp/Relphormer
https://github.com/osome-iu/Rematch-RARE

**Reply by samlhuillier** [reply]
Yes I think this feature will be coming in two stages: first we'll implement manual tagging and then automatic tagging. I think I can give you the estimate that this'll probably come in the next month or so.

In the meantime, the features on linking are now implemented. Have you tried them? I'd love to hear what you think!!
![ezgif-2-b38c66c33a](https://github.com/reorproject/reor/assets/17236551/af668b19-9914-4d56-9409-4e292df5eb81)

**Reply by RazyRo** [reply]
That's fantastic! I had not updated Reor before, and it is great.As for your quick response and work, I can't wait to see this on the knowledge graph. Thank you for looking into our request. it means a lot to us

**Reply by RazyRo** [reply]
this thread was created two months ago, and you mentioned that automatic tagging would likely be implemented by the following month. Could you please provide an update on this feature? I have been eagerly waiting and currently rely on Reor for automatic tagging and the knowledge graph.

**Reply by samlhuillier** [reply]
Sorry haven't been able to get to this! Can't make any promises but am looking into things on topic modelling and have some thoughts about potentially applying it here...So in the coming weeks hopefully something comes of this :)

**Reply by BradKML** [reply]
I am gonna say it: Siyuan is displacing LogSeq as the de-facto FOSS tool (but they have yet to copy Obsidian's agentic note tooling), but everything from LogSeq to Obsidian to Siyuan all have a "graph view"
- https://github.com/glowingjade/obsidian-smart-composer
- https://github.com/logancyang/obsidian-copilot
- https://github.com/brianpetro/obsidian-smart-connections

**Comment by bajman** [comment]
This would be an awesome idea!

**Reply by suhaakan** [reply]
Creating an Obsidian-type graph map manually is a valuable feature, but creating a knowledge graph by llm according to the similarities between our notes will provide a very important insight for the obsidian-style garph map. Because we can miss some connections between notes while manually linking in obsidian application. 

Can't this feature be provided with the repos that @RazyRo  gave as an example?
https://github.com/rahulnyk/knowledge_graph
https://github.com/Alibaba-NLP/KB-NER
https://github.com/zjunlp/Relphormer
https://github.com/osome-iu/Rematch-RARE

**Reply by RazyRo** [reply]
https://github.com/microsoft/graphrag

**Comment by BradKML** [comment]
Would love to see this since LogSeq already pioneered this in FOSS DigiGarden/BASB/Zettlekasten (Obsidian and other did it in proprietary systems)
